pandas/tests/indexes/multi/test_uniq_dups.py

```diff
@@ -0,0 +1,236 @@
+    result = idx.unique(level=level)
+    expected = idx.get_level_values(level).unique()
+    tm.assert_index_equal(result, expected)
+
+    # With already unique level
+    mi = pd.MultiIndex.from_arrays([[1, 3, 2, 4], [1, 3, 2, 5]],
+                                   names=['first', 'second'])
+    result = mi.unique(level=level)
+    expected = mi.get_level_values(level)
+    tm.assert_index_equal(result, expected)
+
+    # With empty MI
+    mi = pd.MultiIndex.from_arrays([[], []], names=['first', 'second'])
+    result = mi.unique(level=level)
+    expected = mi.get_level_values(level)
+
+
+def test_duplicate_multiindex_labels():
+    # GH 17464
+    # Make sure that a MultiIndex with duplicate levels throws a ValueError
+    with pytest.raises(ValueError):
+        ind = pd.MultiIndex([['A'] * 10, range(10)], [[0] * 10, range(10)])
+
+    # And that using set_levels with duplicate levels fails
+    ind = MultiIndex.from_arrays([['A', 'A', 'B', 'B', 'B'],
+                                  [1, 2, 1, 2, 3]])
+    with pytest.raises(ValueError):
+        ind.set_levels([['A', 'B', 'A', 'A', 'B'], [2, 1, 3, -2, 5]],
+                       inplace=True)
+
+
+@pytest.mark.parametrize('names', [['a', 'b', 'a'], ['1', '1', '2'],
+                                   ['1', 'a', '1']])
+def test_duplicate_level_names(names):
+    # GH18872
+    pytest.raises(ValueError, pd.MultiIndex.from_product,
+                  [[0, 1]] * 3, names=names)
+
+    # With .rename()
+    mi = pd.MultiIndex.from_product([[0, 1]] * 3)
+    tm.assert_raises_regex(ValueError, "Duplicated level name:",
+                           mi.rename, names)
+
+    # With .rename(., level=)
+    mi.rename(names[0], level=1, inplace=True)
+    tm.assert_raises_regex(ValueError, "Duplicated level name:",
+                           mi.rename, names[:2], level=[0, 2])
+
+
+def test_duplicate_meta_data():
+    # GH 10115
+    index = MultiIndex(
+        levels=[[0, 1], [0, 1, 2]],
+        labels=[[0, 0, 0, 0, 1, 1, 1],
+                [0, 1, 2, 0, 0, 1, 2]])
+
+    for idx in [index,
+                index.set_names([None, None]),
+                index.set_names([None, 'Num']),
+                index.set_names(['Upper', 'Num']), ]:
+        assert idx.has_duplicates
+        assert idx.drop_duplicates().names == idx.names
+
+
+def test_duplicates(idx):
+    assert not idx.has_duplicates
+    assert idx.append(idx).has_duplicates
+
+    index = MultiIndex(levels=[[0, 1], [0, 1, 2]], labels=[
+        [0, 0, 0, 0, 1, 1, 1], [0, 1, 2, 0, 0, 1, 2]])
+    assert index.has_duplicates
+
+    # GH 9075
+    t = [(u('x'), u('out'), u('z'), 5, u('y'), u('in'), u('z'), 169),
+         (u('x'), u('out'), u('z'), 7, u('y'), u('in'), u('z'), 119),
+         (u('x'), u('out'), u('z'), 9, u('y'), u('in'), u('z'), 135),
+         (u('x'), u('out'), u('z'), 13, u('y'), u('in'), u('z'), 145),
+         (u('x'), u('out'), u('z'), 14, u('y'), u('in'), u('z'), 158),
```

Comment:
just rename module to: test_unique_and_duplicates

Reply:
Done